doc/decisions/map_collaboration_data_synchronization.md

```diff
@@ -0,0 +1,64 @@
+# Map Collaboration & Data Synchronization (Offline)
+
+## Problem
+
+The user wants to view changes that other users have made on the map, therefore the data needs to be kept in sync.
+The user should also be able to access some features of the application while being offline and later synchronize her changes to the server.
+
+## Constraints
+
+- The solution should be simple.
+- If any conflicts arise that can not automatically be resolved, the user should have choice over how to resolve such changes (keep local/keep remote).
+- The solution should only disturb the user experience when necessary.
+- After the conflict is resolved all users should have a consistent state.
+- Users can not concurrently add changes to the same layer (removes complexity).
+
+## Assumptions
+
+- A 100% multiplayer solution is not required, therefore the data can possibly be stale for a short while for some of the users.
+- Collaboration is more important than the feature being available in an offline environment.
+
+## Solutions
+
+### Layer Locking, Version Synchronization.
+
+#### Layer Locking:
+
+If a user wants to change a layer, the layer is locked for the editing user.
+Users without the lock have read-only access to the layer.
+The lock is kept for a certain amount of time but can be refreshed.
+For refreshing a timestamp is kept when the lock got attained/refreshed.
+Users can give up on the lock.
+
+#### Data Synchronisation via Timestamps:
+
+Each action is associated with a timestamp (for ease of implementation this timestamp is the local time in UTC).
+Each object on the map has an associated timestamp as well.
+
+#### Version Synchronisation for Read-only Users:
+
+The data visible to read-only users, is the latest version of the map.
+This data is kept in sync via a polling mechanism.
+
+If the user keeping the lock saves the changes, all built-up changes are bundled into a new version and saved to the database.
```

Comment:
We would then have to keep track of the undo/redo history for each user on the backend as well.
As I understand versioning it is a snapshot of the map in all layers.

The read-only users should only have the latest version of the data + their local changes in the history.

Reply:
> We would then have to keep track of the undo/redo history for each user on the backend as well.

I understand why it could be desirable to do so (to have undo beyond the current session) but I don't understand why it would be needed. Why not only track what happened in the frontend, and when pressing undo, the backend simply gets updated again with the new information, e.g., let us assume a plant is positioned at 10, 10:

action         | API call   | DB | State
-----------------|-------------|----------|----|
 |  | | {10,10} | {**{10, 10}**} |
move {20, 20} | move {20,20}  |  {20,20} | {{10, 10}, **{20,20}**} |
undo | move {10,10}  |  {10,10} | {**{10, 10}**, {20,20}} |
redo | move {20,20}  |  {20,20} | {{10, 10}, **{20,20}**} |

(DB only in backend, State only in frontend)

Offline feature simply would be to send all states to the backend that were not send yet (do all API calls that queued up).

Am I missing something?


> As I understand versioning it is a snapshot of the map in all layers.

Yes, this should be unrelated to everything else and is not a very important feature (a bit more user friendly to use than to duplicate the maps). We can drop it if it interferes with another feature or is otherwise difficult to implement. The feature probably only makes sense if we have the whole change history in the backend.

> The read-only users should only have the latest version of the data + their local changes in the history.

Yes, that would be an alternative but it basically would be a reimplementation of what currently undo does? What I propose seems to me like an extension (send already stored data also to the backend).

> Without keeping track of the undo/redo history on the backend, sending a new version to the backend at a fast pace would clutter it.

What does the last **it** refer to?

> The undo/redo feature would then also only work in a very limited way.

Which limitations? (We can also discuss it today.)